File: sdks/go/reproit/semantic_dependency.go
package reproit

import (
	"encoding/base64"
	"encoding/json"
)

type semanticDependencyMetadata struct {
	Name  []byte
	Value []byte
}

type semanticDependencyRequest struct {
	Encoding         string
	Metadata         []semanticDependencyMetadata
	Method           *string
	ObservationClass automaticObservationClass
	Operation        string
	Payload          []byte
	Protocol         string
	Target           string
}

type semanticDependencyResponse struct {
	ErrorCode   *string
	ErrorNumber *uint32
	Metadata    []semanticDependencyMetadata
	Outcome     observationOutcome
	Payload     []byte
	HasPayload  bool
	Status      *string
	StatusCode  *uint16
}

type sdkEngineDependencyMetadata struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

type sdkEngineDependencyRequest struct {
	Encoding         string                        `json:"encoding"`
	Metadata         []sdkEngineDependencyMetadata `json:"metadata"`
	Method           *string                       `json:"method"`
	ObservationClass string                        `json:"observation_class"`
	Operation        string                        `json:"operation"`
	Payload          string                        `json:"payload"`
	Protocol         string                        `json:"protocol"`
	Target           string                        `json:"target"`
}

type sdkEngineDependencyResponse struct {
	ErrorCode   *string                       `json:"error_code"`
	ErrorNumber *uint32                       `json:"error_number"`
	Metadata    []sdkEngineDependencyMetadata `json:"metadata"`
	Outcome     string                        `json:"outcome"`
	Payload     *string                       `json:"payload"`
	Status      *string                       `json:"status"`
	StatusCode  *uint16                       `json:"status_code"`
}

func translateSemanticDependency(
	operation *AutomaticOperation,
	request semanticDependencyRequest,
	causalParentID *string,
	live func() (semanticDependencyResponse, error),
) (semanticDependencyResponse, error) {
	requestInput, err := makeSDKEngineDependencyRequest(request)
	if err != nil {
		return live()
	}
	started, err := operation.openSemanticDependency(requestInput, causalParentID)
	if err != nil {
		return live()
	}
	finished := false
	defer func() {
		if !finished {
			_ = operation.project.bridge.abandonObservation(
				sdkEngineObservationHandle(started.Handle),
			)
		}
	}()
	if started.Action == "capture" {
		response, liveErr := live()
		responseInput, convertErr := makeSDKEngineDependencyResponse(response)
		if convertErr == nil {
			outcome, finishErr := operation.project.bridge.finishDependency(
				started.Handle, &responseInput,
			)
			finished = finishErr == nil && outcome == string(response.Outcome)
		}
		if !finished {
			_ = operation.markUnowned(request.ObservationClass, causalParentID, nil)
		}
		return response, liveErr
	}
	record, err := readSDKEngineDependencyResponse(operation.project.bridge, started.Handle)
	if err != nil {
		return semanticDependencyResponse{}, ErrAutomaticCapture
	}
	outcome, err := operation.project.bridge.finishDependency(started.Handle, nil)
	if err != nil {
		return semanticDependencyResponse{}, ErrAutomaticCapture
	}
	response, err := reconstructSemanticDependencyResponse(record, outcome)
	if err != nil {
		return semanticDependencyResponse{}, ErrAutomaticCapture
	}
	finished = true
	return response, nil
}

func (operation *AutomaticOperation) openSemanticDependency(
	request sdkEngineDependencyRequest,
	causalParentID *string,
) (sdkEngineDependencyStart, error) {
	operation.mu.Lock()
	defer operation.mu.Unlock()
	if operation.finished || operation.worldComplete {
		return sdkEngineDependencyStart{}, ErrAutomaticCapture
	}
	started, err := operation.project.bridge.openDependency(
		operation.handle, causalParentID, request,
	)
	if err != nil {
		operation.abandonLocked()
		return sdkEngineDependencyStart{}, ErrAutomaticCapture
	}
	return started, nil
}

func makeSDKEngineDependencyRequest(
	request semanticDependencyRequest,
) (sdkEngineDependencyRequest, error) {
	metadata, err := makeSDKEngineDependencyMetadata(request.Metadata)
	if err != nil || !boundedDependencyBytes(
		len(request.Encoding), len(request.Operation), len(request.Payload),
		len(request.Protocol), len(request.Target),
	) {
		return sdkEngineDependencyRequest{}, ErrAutomaticCapture
	}
	return sdkEngineDependencyRequest{
		Encoding:         request.Encoding,
		Metadata:         metadata,
		Method:           request.Method,
		ObservationClass: string(request.ObservationClass),
		Operation:        request.Operation,
		Payload:          base64.RawURLEncoding.EncodeToString(request.Payload),
		Protocol:         request.Protocol,
		Target:           base64.RawURLEncoding.EncodeToString([]byte(request.Target)),
	}, nil
}

func makeSDKEngineDependencyResponse(
	response semanticDependencyResponse,
) (sdkEngineDependencyResponse, error) {
	metadata, err := makeSDKEngineDependencyMetadata(response.Metadata)
	if err != nil || !boundedDependencyBytes(len(response.Payload)) {
		return sdkEngineDependencyResponse{}, ErrAutomaticCapture
	}
	var payload *string
	if response.HasPayload {
		encoded := base64.RawURLEncoding.EncodeToString(response.Payload)
		payload = &encoded
	}
	return sdkEngineDependencyResponse{
		ErrorCode:   response.ErrorCode,
		ErrorNumber: response.ErrorNumber,
		Metadata:    metadata,
		Outcome:     string(response.Outcome),
		Payload:     payload,
		Status:      response.Status,
		StatusCode:  response.StatusCode,
	}, nil
}

func makeSDKEngineDependencyMetadata(
	metadata []semanticDependencyMetadata,
) ([]sdkEngineDependencyMetadata, error) {
	if len(metadata) > sdkEngineMaxCallBytes {
		return nil, ErrAutomaticCapture
	}
	totalBytes := 0
	result := make([]sdkEngineDependencyMetadata, 0, len(metadata))
	for _, field := range metadata {
		if len(field.Name) > sdkEngineMaxCallBytes-totalBytes {
			return nil, ErrAutomaticCapture
		}
		totalBytes += len(field.Name)
		if len(field.Value) > sdkEngineMaxCallBytes-totalBytes {
			return nil, ErrAutomaticCapture
		}
		totalBytes += len(field.Value)
		result = append(result, sdkEngineDependencyMetadata{
			Name:  base64.RawURLEncoding.EncodeToString(field.Name),
			Value: base64.RawURLEncoding.EncodeToString(field.Value),
		})
	}
	return result, nil
}

func readSDKEngineDependencyResponse(
	bridge *sdkEngineBridge,
	handle sdkEngineDependencyHandle,
) ([]byte, error) {
	result := make([]byte, 0, sdkEngineMaxObservationReadBytes)
	for {
		chunk, err := bridge.readObservation(sdkEngineObservationHandle(handle))
		if err != nil || (!chunk.EOF && len(chunk.Chunk) == 0) ||
			len(chunk.Chunk) > sdkEngineMaxSemanticDependencyRecordBytes-len(result) {
			return nil, ErrAutomaticCapture
		}
		result = append(result, chunk.Chunk...)
		if chunk.EOF {
			if len(result) == 0 {
				return nil, ErrAutomaticCapture
			}
			return result, nil
		}
	}
}

func reconstructSemanticDependencyResponse(
	record []byte,
	validatedOutcome string,
) (semanticDependencyResponse, error) {
	var wire sdkEngineDependencyResponse
	if json.Unmarshal(record, &wire) != nil || wire.Outcome != validatedOutcome {
		return semanticDependencyResponse{}, ErrAutomaticCapture
	}
	metadata := make([]semanticDependencyMetadata, 0, len(wire.Metadata))
	for _, field := range wire.Metadata {
		name, nameErr := base64.RawURLEncoding.DecodeString(field.Name)
		value, valueErr := base64.RawURLEncoding.DecodeString(field.Value)
		if nameErr != nil || valueErr != nil {
			return semanticDependencyResponse{}, ErrAutomaticCapture
		}
		metadata = append(metadata, semanticDependencyMetadata{Name: name, Value: value})
	}
	response := semanticDependencyResponse{
		ErrorCode:   wire.ErrorCode,
		ErrorNumber: wire.ErrorNumber,
		Metadata:    metadata,
		Outcome:     observationOutcome(wire.Outcome),
		Status:      wire.Status,
		StatusCode:  wire.StatusCode,
	}
	if wire.Payload != nil {
		payload, err := base64.RawURLEncoding.DecodeString(*wire.Payload)
		if err != nil {
			return semanticDependencyResponse{}, ErrAutomaticCapture
		}
		response.Payload = payload
		response.HasPayload = true
	}
	return response, nil
}

func boundedDependencyBytes(values ...int) bool {
	total := 0
	for _, value := range values {
		if value < 0 || value > sdkEngineMaxCallBytes-total {
			return false
		}
		total += value
	}
	return true
}

File: sdks/dotnet/ReproIt.Sdk/SemanticDependency.cs
using System.Runtime.ExceptionServices;
using System.Text;
using System.Text.Json.Nodes;

namespace ReproIt.Sdk;

internal sealed record SemanticDependencyMetadata(byte[] Name, byte[] Value);

internal sealed record SemanticDependencyRequest(
    string Encoding,
    IReadOnlyList<SemanticDependencyMetadata> Metadata,
    string? Method,
    AutomaticObservationClass ObservationClass,
    string Operation,
    byte[] Payload,
    string Protocol,
    string Target);

internal sealed record SemanticDependencyResponse(
    string? ErrorCode,
    uint? ErrorNumber,
    IReadOnlyList<SemanticDependencyMetadata> Metadata,
    ObservationOutcome Outcome,
    byte[]? Payload,
    string? Status,
    ushort? StatusCode);

internal sealed record SemanticDependencyLiveResult(
    SemanticDependencyResponse Response,
    Exception? Error);

internal sealed class SdkEngineDependencyConnection(
    SdkEngineBridge bridge,
    SdkEngineDependencyStart start)
{
    internal string Action => start.Action;

    internal void Abandon()
    {
        try
        {
            bridge.AbandonObservation(new SdkEngineObservationHandle(start.Handle.Value));
        }
        catch (Exception)
        {
            // Capture cleanup must not change application behavior.
        }
    }

    internal SdkEngineObservationChunk Read() =>
        bridge.ReadObservation(new SdkEngineObservationHandle(start.Handle.Value));

    internal string Finish(JsonObject? response) =>
        bridge.FinishDependency(start.Handle, response);
}

internal static class SemanticDependencyTranslator
{
    private static readonly UTF8Encoding StrictUtf8 = new(false, true);

    internal static SemanticDependencyResponse Translate(
        AutomaticOperation operation,
        SemanticDependencyRequest request,
        string? causalParentId,
        Func<SemanticDependencyLiveResult> live)
    {
        JsonObject requestInput;
        try
        {
            requestInput = MakeRequest(request);
        }
        catch (Exception)
        {
            return ReturnLive(live());
        }
        SdkEngineDependencyConnection connection;
        try
        {
            connection = operation.OpenSemanticDependency(requestInput, causalParentId);
        }
        catch (Exception)
        {
            return ReturnLive(live());
        }
        bool finished = false;
        try
        {
            if (connection.Action == "capture")
            {
                SemanticDependencyLiveResult result = live();
                try
                {
                    _ = connection.Finish(MakeResponse(result.Response));
                    finished = true;
                }
                catch (Exception)
                {
                    // Capture failure must not change the live dependency result.
                }
                return ReturnLive(result);
            }
            byte[] record = ReadReplayResponse(connection);
            string validatedOutcome;
            try
            {
                validatedOutcome = connection.Finish(null);
            }
            catch (Exception)
            {
                throw AutomaticProject.CaptureError();
            }
            SemanticDependencyResponse response = ReconstructResponse(
                record, validatedOutcome);
            finished = true;
            return response;
        }
        finally
        {
            if (!finished)
            {
                connection.Abandon();
            }
        }
    }

    internal static JsonObject MakeRequest(SemanticDependencyRequest request)
    {
        int rawBytes = CheckedBytes(
            StrictUtf8.GetByteCount(request.Encoding),
            request.Payload.Length,
            StrictUtf8.GetByteCount(request.Operation),
            StrictUtf8.GetByteCount(request.Protocol),
            StrictUtf8.GetByteCount(request.Target));
        if (rawBytes > SdkEngineBridge.MaxCallBytes)
        {
            throw AutomaticProject.CaptureError();
        }
        return new JsonObject
        {
            ["encoding"] = request.Encoding,
            ["metadata"] = MakeMetadata(request.Metadata),
            ["method"] = request.Method,
            ["observation_class"] = AutomaticOperation.ObservationClass(
                request.ObservationClass),
            ["operation"] = request.Operation,
            ["payload"] = EncodeBase64Url(request.Payload),
            ["protocol"] = request.Protocol,
            ["target"] = EncodeBase64Url(StrictUtf8.GetBytes(request.Target)),
        };
    }

    internal static JsonObject MakeResponse(SemanticDependencyResponse response)
    {
        if (response.Payload is not null && response.Payload.Length > SdkEngineBridge.MaxCallBytes)
        {
            throw AutomaticProject.CaptureError();
        }
        return new JsonObject
        {
            ["error_code"] = response.ErrorCode,
            ["error_number"] = response.ErrorNumber,
            ["metadata"] = MakeMetadata(response.Metadata),
            ["outcome"] = Outcome(response.Outcome),
            ["payload"] = response.Payload is null ? null : EncodeBase64Url(response.Payload),
            ["status"] = response.Status,
            ["status_code"] = response.StatusCode,
        };
    }

    private static JsonArray MakeMetadata(IReadOnlyList<SemanticDependencyMetadata> metadata)
    {
        if (metadata.Count > SdkEngineBridge.MaxCallBytes)
        {
            throw AutomaticProject.CaptureError();
        }
        int rawBytes = 0;
        JsonArray result = [];
        foreach (SemanticDependencyMetadata field in metadata)
        {
            rawBytes = CheckedBytes(rawBytes, field.Name.Length, field.Value.Length);
            if (rawBytes > SdkEngineBridge.MaxCallBytes)
            {
                throw AutomaticProject.CaptureError();
            }
            result.Add(new JsonObject
            {
                ["name"] = EncodeBase64Url(field.Name),
                ["value"] = EncodeBase64Url(field.Value),
            });
        }
        return result;
    }

    private static byte[] ReadReplayResponse(SdkEngineDependencyConnection connection)
    {
        using MemoryStream result = new();
        while (true)
        {
            SdkEngineObservationChunk chunk;
            try
            {
                chunk = connection.Read();
            }
            catch (Exception)
            {
                throw AutomaticProject.CaptureError();
            }
            if ((!chunk.Eof && chunk.Chunk.Length == 0) ||
                result.Length > SdkEngineBridge.MaxSemanticDependencyRecordBytes -
                    chunk.Chunk.Length)
            {
                throw AutomaticProject.CaptureError();
            }
            result.Write(chunk.Chunk);
            if (chunk.Eof)
            {
                return result.Length > 0
                    ? result.ToArray()
                    : throw AutomaticProject.CaptureError();
            }
        }
    }

    private static SemanticDependencyResponse ReconstructResponse(
        byte[] record,
        string validatedOutcome)
    {
        JsonObject value;
        try
        {
            value = JsonNode.Parse(record)!.AsObject();
        }
        catch (Exception)
        {
            throw AutomaticProject.CaptureError();
        }
        List<SemanticDependencyMetadata> metadata = [];
        try
        {
            foreach (JsonNode? node in value["metadata"]!.AsArray())
            {
                JsonObject field = node!.AsObject();
                metadata.Add(new SemanticDependencyMetadata(
                    DecodeBase64Url(field["name"]), DecodeBase64Url(field["value"])));
            }
            return new SemanticDependencyResponse(
                OptionalText(value["error_code"]),
                OptionalUInt32(value["error_number"]),
                metadata,
                validatedOutcome == "response"
                    ? ObservationOutcome.Response
                    : ObservationOutcome.Error,
                value["payload"] is null ? null : DecodeBase64Url(value["payload"]),
                OptionalText(value["status"]),
                OptionalUInt16(value["status_code"]));
        }
        catch (Exception)
        {
            throw AutomaticProject.CaptureError();
        }
    }

    private static string? OptionalText(JsonNode? value) =>
        value is null ? null : value.GetValue<string>();

    private static uint? OptionalUInt32(JsonNode? value) =>
        value is null ? null : value.GetValue<uint>();

    private static ushort? OptionalUInt16(JsonNode? value) =>
        value is null ? null : value.GetValue<ushort>();

    private static string Outcome(ObservationOutcome outcome) =>
        outcome == ObservationOutcome.Response ? "response" : "error";

    private static SemanticDependencyResponse ReturnLive(SemanticDependencyLiveResult result)
    {
        if (result.Error is not null)
        {
            ExceptionDispatchInfo.Capture(result.Error).Throw();
        }
        return result.Response;
    }

    private static int CheckedBytes(params int[] values)
    {
        int result = 0;
        foreach (int value in values)
        {
            result = checked(result + value);
        }
        return result;
    }

    private static string EncodeBase64Url(ReadOnlySpan<byte> value) =>
        Convert.ToBase64String(value).TrimEnd('=').Replace('+', '-').Replace('/', '_');

    private static byte[] DecodeBase64Url(JsonNode? value)
    {
        string encoded = value?.GetValue<string>() ?? throw AutomaticProject.CaptureError();
        string standard = encoded.Replace('-', '+').Replace('_', '/');
        standard += new string('=', (4 - standard.Length % 4) % 4);
        return Convert.FromBase64String(standard);
    }
}
